# 라이프 사이클과 에러 핸들링

## 잡과 디퍼드

* 결과가 없는 비동기 함수
    * 로그 기록, 분석 데이터 전송, 완료 여부 모니터링과 같은 백그라운드 작업
* 결과를 반환하는 비동기 함수
    * 비동기 함수를 사용한 정보 반환

### 잡

* fire-and-forget 작업
* 시작되면 예외가 발생하지 않는 한 대기하지 않는다
* `launch()`를 사용해 잡을 생성하는 것이 가장 일반적

```kotlin
// launch함수를 사용
fun main(args: Array<String>) = runBlocking {
        val job = GlobalScope.lanch {
            // Do background task here
        }
    }

// Job()팩토리 함수 사용
fun main(args: Array<String>) = runBlocking {
    val job = Job()
}
```

#### 예외처리

* 잡 내부에서 발생하는 예외는 잡을 생성한 곳까지 전파
* 잡이 완료되기를 기다리지 않아도 발생

#### 라이프 사이클

* New: 존재하지만 아직 실행되지 않은 잡
* Active: 실행 중/일시 중단
* Completed: 완료되어 더이상 실행되지 않는 경우
* Canceling: 실행 중인 잡에서 `cancel()`이 호출되면 취소가 완료될 때 까지 시간이 걸림
* Cancelled: 취소로 인해 실행이 완료된 Job

#### 생성

* `launch()`나 `Job()`을 사용해 생성될 때 자동 시작
* 자동 시작되지않게 하려면 `CoroutineStart.LAZY`

#### 활성

* `start()`
    * Job이 완료될 때까지 기다리지 않고 시작
* `join()`
    * Job이 완료될 때까지 실행을 일시 중단

#### 취소 중

* 취소 요청을 받은 활성 Job은 취소중이라고 하는 스테이징 상태로 들어감
* Job 실행 취소 요청: `cancel()`

#### 취소됨

* 취소 또는 처리되지 않은 예외로 인해 실행이 종료된 상태
* Job이 취소되면 `getCancellationException()`함수를 통해 취소에 대한 정보를 얻을 수 있음
* 취소된 Job과 예외로 인해 실패한 Job을 구별하기 위해 `CoroutineExceptionHandler`를 통한 취소 작업 처리

```kotlin
fun main(args: Array<String>) = runBlocking {
    val exceptionHandler = CoroutineExceptionHandler { _: CoroutineContext, throwable: Throwable ->
        println("Job cancelled due to ${throwable.message}")
    }

    val job = GlobalScope.lanch(exceptionHandler) {
        delay(5000)
    }

    delay(2000)
}
```

* 또는 `invokeOnCompletion()`사용

```kotlin
fun main(args: Array<String>) = runBlocking {
    val job = GlobalScope.lanch {
        delay(5000)
    }.invokeOnCompletion { cause ->
        cause?.let {
            println("Job cancelled due to ${throwable.message}")
        }
    }

    delay(2000)
}
```

#### 완료됨

* 실행이 중지된 Job
* 정상 || 취소 || 예외

#### Job의 현재 상태 확인

* `isActive`
    * 활성 상태 여부. 일시 중지도 true
* `isCompleted`
    * Job이 실행을 완료했는지 여부
* `isCancelled`
    * Job 취소 여부. 취소 요청 시 즉시 true

### 디퍼드

* 결과를 갖는 비동기 작업을 수행
    * Futures 또는 Promises
* 연산이 객체를 반환할 것이며, 객체는 비동기 작업이 완료될 때 까지 비어있을 것
* Job과 라이프 사이클은 비슷함
* 디퍼드를 만드는 방법

```kotlin
fun main(args: Array<String>) = runBlocking {
    val headlinesTask = GlobalScope.async {
        getHeadlines()
    }

    headlinesTask.await()
}

// 또는 CompletableDeferred
val articlesTask = CompletableDeferred<List<Article>>()
```

#### 예외 처리

* 처리되지 않은 예외를 자동으로 전파하지 않는다.
* 실행의 성공 여부 판단은 사용자의 몫
* 예외 전파 방법

```kotlin
fun main(args: Array<String>) = runBlocking<Unit> {
    val deferred = GlobalScope.async {
        TODO("Not implemented yet")
    }

    // 예외 처리
    try {
        deferred.await() // 예외 전파 및 중단
    } catch (throwable: Throwable) {
        println("Deferred cancelled due to ${throwable.message}")
    }
}
```

#### 디퍼드가 예외를 갖도록 하기

* 예외를 처리하는 쉬운 방법은 `await()` 대신 `join()`을 사용해 디퍼드를 기다리는 것
* 이렇게 하면 대기할 때 예외가 전파되지 않음
* `join()`을 사용해 디퍼드가 대기된 경우, 예외가 전파되지 않도록 값을 읽기 전에 취소 여부를 확인해야 함
* Job에 예외를 기록하거나 표시해야..

### 상태는 한 방향으로만 이동

* Job이 특정 상태에 도달하면 이전 상태로 돌아가지 않는다.

```kotlin
fun main(args: Array<String>) = runBlocking {
    val time = measureTimeMills {
        val job = GlobalScope.launch {
            delay(2000)
        }
        // wait for it to complete once
        job.join()
        // Restart the Job
        job.start()
        job.join()
    }
    println("Took $time ms")
}
```

* 이미 완료된 Job에서 `start()`를 호출해 다시 시작했다면 실행 시간은 4초가 될 것
    * But,... Job은 이미 Completed 상태에 도달했으므로 `start()`를 호출해도 아무런 변화가 없음
* 즉 총 실행 시간은 그대로 2초 소요
* 최종 상태는 Job이 이동할 수 없는 상태
    * `Cancelled`
    * `Completed`

